# Used Car Price Prediction

In [1]:
print("Hello world")

Hello world


## Import libraries

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error


### Importing dataset from kaggle directly

In [6]:
from google.colab import files
files.upload()   # upload kaggle.json


Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"mukeshsunku","key":"0ba85e306b541aa18a7549b3a596f382"}'}

In [7]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json


# Basic model with basic dataset

In [8]:
!kaggle datasets download -d mohitkumar282/used-car-dataset

Dataset URL: https://www.kaggle.com/datasets/mohitkumar282/used-car-dataset
License(s): CC0-1.0
  0% 0.00/590k [00:00<?, ?B/s]
100% 590k/590k [00:00<00:00, 1.64GB/s]


In [9]:
!ls

kaggle.json  sample_data  used-car-dataset.zip


In [10]:
!unzip used-car-dataset.zip

Archive:  used-car-dataset.zip
  inflating: used_car_dataset.csv    
  inflating: used_cars_dataset_v2.csv  


# 1


## Explore the dataset

In [137]:
df1 = pd.read_csv('used_car_dataset.csv')
df1

,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,"98,000 km",Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...","₹ 1,95,000"
1,Toyota,Innova,2009,15,190000.0 km,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...","₹ 3,75,000"
2,Volkswagen,VentoTest,2010,14,"77,246 km",Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...","₹ 1,84,999"
3,Maruti Suzuki,Swift,2017,7,"83,500 km",Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,"₹ 5,65,000"
4,Maruti Suzuki,Baleno,2019,5,"45,000 km",Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol","₹ 6,85,000"
...,...,...,...,...,...,...,...,...,...,...,...
9577,Skoda,Octavia,2014,10,"105,904 km",Automatic,second,Diesel,Oct-24,"Skoda Octavia 1.9 Elegance TDI, 2014, Diesel","₹ 10,40,000"
9578,Maruti Suzuki,Alto-800,2020,4,"55,000 km",Manual,first,Hybrid/CNG,Nov-24,"Maruti Suzuki Alto 800 CNG LXI Optional, 2020,...","₹ 3,75,000"
9579,Maruti Suzuki,Ritz,2013,11,"92,000 km",Manual,first,Diesel,Nov-24,"Maruti Suzuki Ritz VDi, 2013, Diesel","₹ 4,15,000"
9580,Hyundai,Verna,2019,5,"72,000 km",Automatic,first,Petrol,Oct-24,"Hyundai Verna VTVT 1.6 AT SX Option, 2019, Petrol","₹ 8,55,000"


In [138]:
df2 = pd.read_csv("used_cars_dataset_v2.csv")
df2

,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,"98,000 km",Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...","₹ 1,95,000"
1,Toyota,Innova,2009,15,190000.0 km,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...","₹ 3,75,000"
2,Volkswagen,VentoTest,2010,14,"77,246 km",Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...","₹ 1,84,999"
3,Maruti Suzuki,Swift,2017,7,"83,500 km",Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,"₹ 5,65,000"
4,Maruti Suzuki,Baleno,2019,5,"45,000 km",Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol","₹ 6,85,000"
...,...,...,...,...,...,...,...,...,...,...,...
14988,Maruti Suzuki,Wagon-R,2022,2,"31,000 km",Manual,first,hybrid,Nov-24,"Maruti Suzuki Wagon R LXI CNG Optional, 2022, ...","₹ 5,75,000"
14989,BMW,X1,2017,7,"45,000 km",Automatic,second,Diesel,Dec-24,"BMW X1 2.0 SDRIVE 20D, 2017, Diesel","₹ 13,45,000"
14990,Maruti Suzuki,New-gen Swift,2011,13,"60,000 km",Automatic,second,Diesel,Nov-24,Bahut badhiya new condition,"₹ 1,60,000"
14991,Tata,Zest,2019,5,"62,000 km",Manual,second,hybrid,Nov-24,"Tata Zest Revotron 1.2T XE, 2019, CNG & Hybrids","₹ 3,80,000"


In [139]:
df1.head()

,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,"98,000 km",Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...","₹ 1,95,000"
1,Toyota,Innova,2009,15,190000.0 km,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...","₹ 3,75,000"
2,Volkswagen,VentoTest,2010,14,"77,246 km",Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...","₹ 1,84,999"
3,Maruti Suzuki,Swift,2017,7,"83,500 km",Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,"₹ 5,65,000"
4,Maruti Suzuki,Baleno,2019,5,"45,000 km",Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol","₹ 6,85,000"


In [140]:
df2.head()

,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,"98,000 km",Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...","₹ 1,95,000"
1,Toyota,Innova,2009,15,190000.0 km,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...","₹ 3,75,000"
2,Volkswagen,VentoTest,2010,14,"77,246 km",Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...","₹ 1,84,999"
3,Maruti Suzuki,Swift,2017,7,"83,500 km",Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,"₹ 5,65,000"
4,Maruti Suzuki,Baleno,2019,5,"45,000 km",Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol","₹ 6,85,000"


In [141]:
print("Old Shape:", df1.shape)
print("V2 Shape:", df2.shape)

print("\nOld Missing:\n", df1.isnull().sum())
print("\nV2 Missing:\n", df2.isnull().sum())

Old Shape: (9582, 11)
V2 Shape: (14993, 11)

Old Missing:
 Brand            0
model            0
Year             0
Age              0
kmDriven        47
Transmission     0
Owner            0
FuelType         0
PostedDate       0
AdditionInfo     0
AskPrice         0
dtype: int64

V2 Missing:
 Brand            0
model            0
Year             0
Age              0
kmDriven        88
Transmission     0
Owner            0
FuelType         0
PostedDate       0
AdditionInfo     0
AskPrice         0
dtype: int64


Now, I am going to use the v2 dataset because it contains more data and with less null values as compared

In [142]:
df = df2.copy()

df.columns

Index(['Brand', 'model', 'Year', 'Age', 'kmDriven', 'Transmission', 'Owner',
       'FuelType', 'PostedDate', 'AdditionInfo', 'AskPrice'],
      dtype='object')

In [143]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14993 entries, 0 to 14992
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Brand         14993 non-null  object
 1   model         14993 non-null  object
 2   Year          14993 non-null  int64 
 3   Age           14993 non-null  int64 
 4   kmDriven      14905 non-null  object
 5   Transmission  14993 non-null  object
 6   Owner         14993 non-null  object
 7   FuelType      14993 non-null  object
 8   PostedDate    14993 non-null  object
 9   AdditionInfo  14993 non-null  object
 10  AskPrice      14993 non-null  object
dtypes: int64(2), object(9)
memory usage: 1.3+ MB


In [144]:
df.describe()

,Year,Age
count,14993.000000,14993.000000
mean,2016.322150,7.677850
std,4.343017,4.343017
min,1900.000000,0.000000
25%,2014.000000,5.000000
50%,2017.000000,7.000000
75%,2019.000000,10.000000
max,2024.000000,124.000000


In [145]:
# df['kmDriven'].fillna(df['kmDriven'].median(), inplace=True)
# # Why median? Less affected by extreme values.

# In this code we get the error

Removing 'km' from the kmdriven column and '$' from askprice column aslo to convert that into interger coloumn

In [146]:
# Remove 'km', commas, and spaces
df['kmDriven'] = df['kmDriven'].str.replace('km', '', regex=False)
df['kmDriven'] = df['kmDriven'].str.replace(',', '', regex=False)
df['kmDriven'] = df['kmDriven'].str.strip()

# Convert to float
df['kmDriven'] = pd.to_numeric(df['kmDriven'], errors='coerce')


In [147]:
# Remove currency symbols and commas
df['AskPrice'] = df['AskPrice'].str.replace('₹', '', regex=False)
df['AskPrice'] = df['AskPrice'].str.replace('INR', '', regex=False)
df['AskPrice'] = df['AskPrice'].str.replace(',', '', regex=False)
df['AskPrice'] = df['AskPrice'].str.strip()

# Convert to numeric
df['AskPrice'] = pd.to_numeric(df['AskPrice'], errors='coerce')


In [148]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14993 entries, 0 to 14992
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Brand         14993 non-null  object 
 1   model         14993 non-null  object 
 2   Year          14993 non-null  int64  
 3   Age           14993 non-null  int64  
 4   kmDriven      14905 non-null  float64
 5   Transmission  14993 non-null  object 
 6   Owner         14993 non-null  object 
 7   FuelType      14993 non-null  object 
 8   PostedDate    14993 non-null  object 
 9   AdditionInfo  14993 non-null  object 
 10  AskPrice      14993 non-null  int64  
dtypes: float64(1), int64(3), object(7)
memory usage: 1.3+ MB


In [149]:
df.isnull().sum()

,0
Brand,0
model,0
Year,0
Age,0
kmDriven,88
Transmission,0
Owner,0
FuelType,0
PostedDate,0
AdditionInfo,0


In [150]:
df.head()

,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,98000.0,Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...",195000
1,Toyota,Innova,2009,15,190000.0,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...",375000
2,Volkswagen,VentoTest,2010,14,77246.0,Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...",184999
3,Maruti Suzuki,Swift,2017,7,83500.0,Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,565000
4,Maruti Suzuki,Baleno,2019,5,45000.0,Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol",685000


In [151]:
df['kmDriven'].fillna(df['kmDriven'].median(), inplace=True)



/tmp/ipython-input-1460846595.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['kmDriven'].fillna(df['kmDriven'].median(), inplace=True)


In [152]:
df.isnull().sum()

,0
Brand,0
model,0
Year,0
Age,0
kmDriven,0
Transmission,0
Owner,0
FuelType,0
PostedDate,0
AdditionInfo,0


Now the dataset is cleaned

In [153]:
# In the given dataset there are the two columns year and age are both used for same.
# So using both may confuse model. We will keep only one (better: Age).
df.drop('Year', axis=1, inplace=True)


In [154]:
# these are the unwanted columns in the feature selection, so we are deleting those columns
df.drop(['model', 'PostedDate', 'AdditionInfo'], axis=1, inplace=True)


“I performed feature selection by removing high-cardinality and noisy attributes such as model names, posting dates, and unstructured text fields.”

In [155]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14993 entries, 0 to 14992
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Brand         14993 non-null  object 
 1   Age           14993 non-null  int64  
 2   kmDriven      14993 non-null  float64
 3   Transmission  14993 non-null  object 
 4   Owner         14993 non-null  object 
 5   FuelType      14993 non-null  object 
 6   AskPrice      14993 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 820.1+ KB


In [156]:
df.describe()

,Age,kmDriven,AskPrice
count,14993.000000,14993.000000,1.499300e+04
mean,7.677850,71043.845194,9.970175e+05
std,4.343017,57291.163563,1.586640e+06
min,0.000000,0.000000,1.500000e+04
25%,5.000000,44500.000000,3.490000e+05
50%,7.000000,65200.000000,5.700000e+05
75%,10.000000,86000.000000,1.025000e+06
max,124.000000,980002.000000,4.250000e+07


In [157]:
df.value_counts()

Brand       Age  kmDriven  Transmission  Owner   FuelType  AskPrice
BMW         8    47000.0   Automatic     second  Diesel    2500000     26
            9    61000.0   Automatic     second  Diesel    1500000     25
Audi        9    42000.0   Automatic     second  Diesel    1050000     23
            10   45000.0   Automatic     second  Diesel    1200000     22
Volkswagen  6    52000.0   Automatic     first   Diesel    1915000     14
                                                                       ..
Jeep        6    40000.0   Automatic     second  Petrol    1150000      1
                 46300.0   Manual        second  Diesel    990000       1
                 47000.0   Manual        second  Diesel    1050000      1
                 52000.0   Manual        first   Diesel    1049000      1
            5    73000.0   Manual        first   Diesel    1175000      1
Name: count, Length: 13809, dtype: int64

In [158]:
df['Brand'].nunique()

43

In [159]:
df['Brand'].unique()

array(['Honda', 'Toyota', 'Volkswagen', 'Maruti Suzuki', 'BMW', 'Ford',
       'Kia', 'Mercedes-Benz', 'Hyundai', 'Audi', 'Renault', 'MG',
       'Volvo', 'Skoda', 'Tata', 'Mahindra', 'Mini', 'Land Rover', 'Jeep',
       'Chevrolet', 'Jaguar', 'Fiat', 'Aston Martin', 'Porsche', 'Nissan',
       'Force', 'Mitsubishi', 'Lexus', 'Isuzu', 'Datsun', 'Ambassador',
       'Rolls-Royce', 'ICML', 'Bajaj', 'Opel', 'Ashok', 'Bentley',
       'Ssangyong', 'Maserati', 'Toyota Land', 'Citroen', 'Lamborghini',
       'Hummer'], dtype=object)

## Encode Categorical Columns

ML models cannot understand text:

Maruti, Manual, Petrol etc., We must convert them to numbers.
We’ll use One-Hot Encoding

In [160]:
df.shape

(14993, 7)

In [161]:
X = df.drop('AskPrice', axis=1)
y = df['AskPrice']


In [162]:
X = pd.get_dummies(X, columns=[
    'Brand',
    'Transmission',
    'Owner',
    'FuelType'
], drop_first=True)


In [163]:
print(X.shape)


(14993, 49)


In [164]:
X.head()

,Age,kmDriven,Brand_Ashok,Brand_Aston Martin,Brand_Audi,Brand_BMW,Brand_Bajaj,Brand_Bentley,Brand_Chevrolet,Brand_Citroen,...,Brand_Tata,Brand_Toyota,Brand_Toyota Land,Brand_Volkswagen,Brand_Volvo,Transmission_Manual,Owner_second,FuelType_Hybrid/CNG,FuelType_Petrol,FuelType_hybrid
0,23,98000.0,False,False,False,False,False,False,False,False,...,False,False,False,False,False,True,True,False,True,False
1,15,190000.0,False,False,False,False,False,False,False,False,...,False,True,False,False,False,True,True,False,False,False
2,14,77246.0,False,False,False,False,False,False,False,False,...,False,False,False,True,False,True,False,False,False,False
3,7,83500.0,False,False,False,False,False,False,False,False,...,False,False,False,False,False,True,True,False,False,False
4,5,45000.0,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,True,False


After get_dummies():

Each category became its own column.

Petrol → FuelType_Petrol

Diesel → FuelType_Diesel

So 1 column → many binary columns.



## Training the Linear Regression model

### Test train split

In [165]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)


In [166]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)


LinearRegression()

In [167]:
# Evalute

from sklearn.metrics import r2_score, mean_squared_error
import numpy as np

y_pred = model.predict(X_test)


print("R2 Score:", r2_score(y_test, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred)))


R2 Score: 0.4988964634651467
RMSE: 1006207.092595793


In [168]:
r2_lr = r2_score(y_test, y_pred)
rmse_lr = np.sqrt(mean_squared_error(y_test, y_pred))

This model given the less performance, now we go through the random forest for better performance of the model

## Random forest

In [169]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=200,
    random_state=42,
    n_jobs=-1
)

rf.fit(X_train, y_train)


RandomForestRegressor(n_estimators=200, n_jobs=-1, random_state=42)

In [170]:
y_pred_rf = rf.predict(X_test)

print("R2 Score (RF):", r2_score(y_test, y_pred_rf))
print("RMSE (RF):", np.sqrt(mean_squared_error(y_test, y_pred_rf)))


R2 Score (RF): 0.6214405371316805
RMSE (RF): 874562.1823773655


In [171]:
r2_rf = r2_score(y_test, y_pred_rf)
rmse_rf = np.sqrt(mean_squared_error(y_test, y_pred_rf))

### Tuned RF

In [172]:
rf_tuned = RandomForestRegressor(
    n_estimators=400,
    max_depth=20,
    min_samples_split=5,
    min_samples_leaf=2,
    random_state=42,
    n_jobs=-1
)

rf_tuned.fit(X_train, y_train)


RandomForestRegressor(max_depth=20, min_samples_leaf=2, min_samples_split=5,
                      n_estimators=400, n_jobs=-1, random_state=42)

In [173]:
y_pred_tuned = rf_tuned.predict(X_test)

print("R2 Score (Tuned RF):", r2_score(y_test, y_pred_tuned))
print("RMSE (Tuned RF):", np.sqrt(mean_squared_error(y_test, y_pred_tuned)))


R2 Score (Tuned RF): 0.6671534141221599
RMSE (Tuned RF): 820060.122391706


In [174]:
r2_rf_tuned = r2_score(y_test, y_pred_tuned)
rmse_rf_tuned = np.sqrt(mean_squared_error(y_test, y_pred_tuned))

## Gradient boosting

In [175]:
from sklearn.ensemble import GradientBoostingRegressor

gbr = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=4,
    random_state=42
)

gbr.fit(X_train, y_train)

y_pred_gbr = gbr.predict(X_test)

print("R2 (GBR):", r2_score(y_test, y_pred_gbr))
print("RMSE (GBR):", np.sqrt(mean_squared_error(y_test, y_pred_gbr)))


R2 (GBR): 0.5792860949232393
RMSE (GBR): 921970.6022874538


In [176]:
r2_gbr = r2_score(y_test, y_pred_gbr)
rmse_gbr = np.sqrt(mean_squared_error(y_test, y_pred_gbr))

“After experimenting with multiple regression models, Random Forest with hyperparameter tuning achieved the best performance (R² ≈ 0.67).”

In [177]:
import pandas as pd

importance = rf_tuned.feature_importances_
features = X.columns

imp_df = pd.DataFrame({
    'Feature': features,
    'Importance': importance
}).sort_values(by='Importance', ascending=False)

imp_df.head(10)


,Feature,Importance
0,Age,0.256007
1,kmDriven,0.201191
29,Brand_Mercedes-Benz,0.114710
23,Brand_Land Rover,0.070586
5,Brand_BMW,0.060094
34,Brand_Porsche,0.052182
40,Brand_Toyota,0.045124
24,Brand_Lexus,0.033672
47,FuelType_Petrol,0.029257
44,Transmission_Manual,0.026052


**Conclusion:**

  In this project, a machine learning model was developed to predict used car prices based on features such as brand, age, kilometers driven, fuel type, transmission, and ownership details.

The dataset was preprocessed by handling missing values, converting categorical variables using one-hot encoding, and removing irrelevant attributes. Multiple regression models including Linear Regression, Random Forest, and Gradient Boosting were implemented and evaluated.

Among the tested models, the tuned Random Forest Regressor achieved the best performance with an R² score of approximately 0.67 and an RMSE of about ₹8.2 lakhs. This indicates that the model was able to explain a significant portion of price variation in the dataset.

The performance was limited by the absence of important vehicle specifications such as engine capacity, mileage, and power. Future improvements can be achieved by using richer datasets and advanced ensemble learning techniques.

Overall, this project provided practical experience in data preprocessing, feature engineering, model selection, and performance evaluation for real-world regression problems.

In the previous model performance is not good. So, now I chose the cardekho dataset for the better performance to the model.

# Advanced car price prediction using cardekho dataset

In [68]:
!kaggle datasets download -d manishkr1754/cardekho-used-car-data

Dataset URL: https://www.kaggle.com/datasets/manishkr1754/cardekho-used-car-data
License(s): DbCL-1.0
  0% 0.00/230k [00:00<?, ?B/s]
100% 230k/230k [00:00<00:00, 633MB/s]


In [69]:
!unzip cardekho-used-car-data.zip

Archive:  cardekho-used-car-data.zip
  inflating: cardekho_dataset.csv    


#2

## Dataset exploring and cleaning

In [178]:
df3 = pd.read_csv('cardekho_dataset.csv')
df3

,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15406,19537,Hyundai i10,Hyundai,i10,9,10723,Dealer,Petrol,Manual,19.81,1086,68.05,5,250000
15407,19540,Maruti Ertiga,Maruti,Ertiga,2,18000,Dealer,Petrol,Manual,17.50,1373,91.10,7,925000
15408,19541,Skoda Rapid,Skoda,Rapid,6,67000,Dealer,Diesel,Manual,21.14,1498,103.52,5,425000
15409,19542,Mahindra XUV500,Mahindra,XUV500,5,3800000,Dealer,Diesel,Manual,16.00,2179,140.00,7,1225000


In [179]:
df3.head()

,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000


In [180]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15411 entries, 0 to 15410
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         15411 non-null  int64  
 1   car_name           15411 non-null  object 
 2   brand              15411 non-null  object 
 3   model              15411 non-null  object 
 4   vehicle_age        15411 non-null  int64  
 5   km_driven          15411 non-null  int64  
 6   seller_type        15411 non-null  object 
 7   fuel_type          15411 non-null  object 
 8   transmission_type  15411 non-null  object 
 9   mileage            15411 non-null  float64
 10  engine             15411 non-null  int64  
 11  max_power          15411 non-null  float64
 12  seats              15411 non-null  int64  
 13  selling_price      15411 non-null  int64  
dtypes: float64(2), int64(6), object(6)
memory usage: 1.6+ MB


In [181]:
df3.isnull().sum()

,0
Unnamed: 0,0
car_name,0
brand,0
model,0
vehicle_age,0
km_driven,0
seller_type,0
fuel_type,0
transmission_type,0
mileage,0


In [182]:
df3.describe()

,Unnamed: 0,vehicle_age,km_driven,mileage,engine,max_power,seats,selling_price
count,15411.000000,15411.000000,1.541100e+04,15411.000000,15411.000000,15411.000000,15411.000000,1.541100e+04
mean,9811.857699,6.036338,5.561648e+04,19.701151,1486.057751,100.588254,5.325482,7.749711e+05
std,5643.418542,3.013291,5.161855e+04,4.171265,521.106696,42.972979,0.807628,8.941284e+05
min,0.000000,0.000000,1.000000e+02,4.000000,793.000000,38.400000,0.000000,4.000000e+04
25%,4906.500000,4.000000,3.000000e+04,17.000000,1197.000000,74.000000,5.000000,3.850000e+05
50%,9872.000000,6.000000,5.000000e+04,19.670000,1248.000000,88.500000,5.000000,5.560000e+05
75%,14668.500000,8.000000,7.000000e+04,22.700000,1582.000000,117.300000,5.000000,8.250000e+05
max,19543.000000,29.000000,3.800000e+06,33.540000,6592.000000,626.000000,9.000000,3.950000e+07


In [183]:
df3.shape

(15411, 14)

In [184]:
df3 = df3.drop('Unnamed: 0', axis=1) #index column is unneccesary so removing


In [185]:
# car_name = brand + model; so removing the car name column for reduce the redundancy
df3 = df3.drop('car_name', axis=1)


## Feature Engineering (Boost Accuracy)

In [186]:
df3.columns

Index(['brand', 'model', 'vehicle_age', 'km_driven', 'seller_type',
       'fuel_type', 'transmission_type', 'mileage', 'engine', 'max_power',
       'seats', 'selling_price'],
      dtype='object')

In [187]:
# creating smart features

# KM Per Year (Usage Intensity)

df3['km_per_year'] = df3['km_driven'] / (df3['vehicle_age'] + 1)


In [188]:
# Power-to-Engine Ratio (Performance)
df3['power_to_engine'] = df3['max_power'] / df3['engine']

In [189]:
# Luxury Brand Flag
luxury_brands = ['BMW','Audi','Mercedes','Jaguar','Volvo','Lexus','Land Rover']

df3['is_luxury'] = df3['brand'].isin(luxury_brands).astype(int)

In [190]:
df3.columns

Index(['brand', 'model', 'vehicle_age', 'km_driven', 'seller_type',
       'fuel_type', 'transmission_type', 'mileage', 'engine', 'max_power',
       'seats', 'selling_price', 'km_per_year', 'power_to_engine',
       'is_luxury'],
      dtype='object')

In [191]:
df3.shape

(15411, 15)

In [192]:
df3.head()

,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price,km_per_year,power_to_engine,is_luxury
0,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000,12000.000000,0.058166,0
1,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000,3333.333333,0.068505,0
2,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000,5000.000000,0.066834,0
3,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000,3700.000000,0.067234,0
4,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000,4285.714286,0.065814,0


## Separate X and y + Log Transform

In [193]:
# Price is skewed → use log.

import numpy as np

X = df3.drop('selling_price', axis=1)
y = np.log1p(df3['selling_price'])

## Encode Categorical Columns

In [194]:
X = pd.get_dummies(
    X,
    columns=['brand','model','seller_type','fuel_type','transmission_type'],
    drop_first=True
)


## Train-Test Split

In [195]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)


I am not goint to try Linear Regression or Random Forest for this advanced project.

"Since I had already established baselines in my earlier project, I directly used gradient boosting for this feature-rich dataset to maximize performance.”

## XG Boost

In [196]:
from xgboost import XGBRegressor

xgb = XGBRegressor(
    n_estimators=500,
    learning_rate=0.05,
    max_depth=6,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    n_jobs=-1
)

xgb.fit(X_train, y_train)


XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.8, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.05, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=6,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=500,
             n_jobs=-1, num_parallel_tree=None, ...)

In [197]:
from sklearn.metrics import r2_score, mean_squared_error

y_pred_log = xgb.predict(X_test)

y_pred = np.expm1(y_pred_log)
y_true = np.expm1(y_test)

print("R2:", r2_score(y_true, y_pred))
print("RMSE:", np.sqrt(mean_squared_error(y_true, y_pred)))


R2: 0.8856567186836763
RMSE: 293386.33239983657


In [198]:
r2_xgb = r2_score(y_true, y_pred)
rmse_xgb = np.sqrt(mean_squared_error(y_true, y_pred))

#### Try Slightly Deeper Trees for better performance

In [199]:
xgb2 = XGBRegressor(
    n_estimators=700,
    learning_rate=0.03,
    max_depth=8,
    subsample=0.85,
    colsample_bytree=0.85,
    random_state=42,
    n_jobs=-1
)

xgb2.fit(X_train, y_train)


XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.85, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.03, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=8,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=700,
             n_jobs=-1, num_parallel_tree=None, ...)

In [200]:
# # Remove Extreme Outliers
# q1 = df3['selling_price'].quantile(0.01)
# q99 = df3['selling_price'].quantile(0.99)
# df_u = df3[(df3['selling_price'] >= q1) & (df3['selling_price'] <= q99)]

In [201]:
import numpy as np
from sklearn.metrics import r2_score, mean_squared_error

# Predict in log scale
y_pred_log = xgb2.predict(X_test)

# Convert back to original price scale
y_pred = np.expm1(y_pred_log)
y_true = np.expm1(y_test)

# Calculate metrics
r2 = r2_score(y_true, y_pred)
rmse = np.sqrt(mean_squared_error(y_true, y_pred))

print("R2 (XGB Tuned):", r2)
print("RMSE (XGB Tuned):", rmse)

R2 (XGB Tuned): 0.9149685600586691
RMSE (XGB Tuned): 253002.21265119503


In [202]:
r2_xgb_tuned = r2_score(y_true, y_pred)
rmse_xgb_tuned = np.sqrt(mean_squared_error(y_true, y_pred))

## Model Evaluation and Conclusion


In [204]:
import pandas as pd

performance_table = pd.DataFrame({
    "Model": [
        "Linear Regression",
        "Random Forest",
        "Tuned Random Forest",
        "Gradient Boosting",
        "XGBoost",
        "Tuned XGBoost"
    ],

    "R2 Score": [
        r2_lr,
        r2_rf,
        r2_rf_tuned,
        r2_gbr,
        r2_xgb,
        r2_xgb_tuned
    ],

    "RMSE (₹)": [
        rmse_lr,
        rmse_rf,
        rmse_rf_tuned,
        rmse_gbr,
        rmse_xgb,
        rmse_xgb_tuned
    ]
})

performance_table


,Model,R2 Score,RMSE (₹)
0,Linear Regression,0.498896,1.006207e+06
1,Random Forest,0.621441,8.745622e+05
2,Tuned Random Forest,0.667153,8.200601e+05
3,Gradient Boosting,0.579286,9.219706e+05
4,XGBoost,0.885657,2.933863e+05
5,Tuned XGBoost,0.914969,2.530022e+05


In [206]:
performance_table["R2 Score"] = performance_table["R2 Score"].round(4)
performance_table["RMSE (₹)"] = performance_table["RMSE (₹)"].round(0)

performance_table


,Model,R2 Score,RMSE (₹)
0,Linear Regression,0.4989,1006207.0
1,Random Forest,0.6214,874562.0
2,Tuned Random Forest,0.6672,820060.0
3,Gradient Boosting,0.5793,921971.0
4,XGBoost,0.8857,293386.0
5,Tuned XGBoost,0.9150,253002.0


In [208]:
performance_table.sort_values(by="R2 Score", ascending=False)


,Model,R2 Score,RMSE (₹)
5,Tuned XGBoost,0.9150,253002.0
4,XGBoost,0.8857,293386.0
2,Tuned Random Forest,0.6672,820060.0
1,Random Forest,0.6214,874562.0
3,Gradient Boosting,0.5793,921971.0
0,Linear Regression,0.4989,1006207.0


In this project, an end-to-end machine learning pipeline was developed to predict used car prices using real-world data. The workflow included data cleaning, feature engineering, categorical encoding, and model training using multiple regression algorithms.

Initially, Linear Regression was used as a baseline model, which provided moderate performance. Ensemble-based models such as Random Forest and Gradient Boosting showed improved results by capturing non-linear relationships in the data. Further performance gains were achieved by tuning the hyperparameters of the Random Forest model.

To achieve higher accuracy, XGBoost was implemented along with logarithmic transformation of the target variable and advanced feature engineering. This significantly enhanced the model’s learning capability. After hyperparameter tuning, the Tuned XGBoost Regressor achieved the best performance with an R² score of approximately 91.5% and an RMSE of around ₹2.5 lakhs.

The results demonstrate that gradient boosting techniques combined with high-quality data and proper preprocessing can produce highly accurate predictive models. This project highlights the importance of systematic model experimentation, feature engineering, and optimization in building reliable machine learning systems.

The final tuned XGBoost model was selected for deployment due to its superior accuracy and stability.

In [209]:
# Sort table by best R2 score
sorted_table = performance_table.sort_values(
    by="R2 Score",
    ascending=False
)

# Get best model row
best_row = sorted_table.iloc[0]

best_name = best_row["Model"]
best_r2 = best_row["R2 Score"]
best_rmse = best_row["RMSE (₹)"]

# Generate conclusion text
conclusion = f"""
Conclusion:

Multiple regression models were evaluated for used car price prediction.
Among all tested models, {best_name} achieved the best performance
with an R2 score of {best_r2:.4f} and RMSE of ₹{best_rmse:,.0f}.

Therefore, {best_name} was selected as the final model for deployment.
"""

print(conclusion)



Conclusion:

Multiple regression models were evaluated for used car price prediction.
Among all tested models, Tuned XGBoost achieved the best performance
with an R2 score of 0.9150 and RMSE of ₹253,002.

Therefore, Tuned XGBoost was selected as the final model for deployment.



## Deployment

In [210]:
# Save Your Trained Model
import joblib

joblib.dump(xgb2, "car_price_model.pkl")


['car_price_model.pkl']

In [211]:
# Save Feature Columns (VERY IMPORTANT)
# Because your model expects exact columns.

joblib.dump(X.columns, "model_features.pkl")


['model_features.pkl']